### PR TITLE
enable e2e test cases on vsts

### PR DESCRIPTION
**Pull request by neil-yechenwei**

### CODE REVIEW

File: tests/test_iotedgedev.py [hunk 1]
```diff
@@ -12,24 +13,35 @@
...
 test_solution_dir = os.path.join(tests_dir, test_solution)
 launch_json_file = os.path.join(tests_dir, "assets", "launch.json")
+launch_json_file_without_nodejs = os.path.join(tests_dir, "assets", "launch_without_nodejs.json")
 
 test_solution_shared_lib = "test_solution_shared_lib"
 test_solution_shared_lib_dir = os.path.join(tests_dir, "assets", test_solution_shared_lib)
 
 
-@pytest.fixture(scope="module", autouse=True)
+def get_docker_os_type():
```
Comment (LazarusX) [inline tests/test_iotedgedev.py]: This method is shared by other tests. You can move it to the utility class.

Comment (neil-yechenwei) [reply tests/test_iotedgedev.py]: Updated code.

File: tests/test_simulator.py [hunk 2]
```diff
@@ -44,6 +47,7 @@ def clean():
     return
 
 
+@pytest.mark.skipif(get_docker_client().get_os_type().lower() == 'windows', reason='Simulator does not support windows container')
```
Comment (LazarusX) [inline tests/test_simulator.py]: These `get_docker_client().get_os_type().lower()` code can be extracted to a method.

File: tox.ini [hunk 3]
```diff
@@ -12,15 +12,17 @@ envlist = py27, py36
...
+    -rrequirements_dev.txt
 commands = pytest -s -v
 #setenv =
 #    PYTHONPATH = {toxinidir}
 
 #commands = 
 #    python setup.py test
 
-passenv = APPDATA ProgramFiles USERPROFILE PROGRAMDATA DOTENV_FILE LOCALAPPDATA DOTNET_CLI_HOME HOME
+passenv = *
```
Comment (LazarusX) [inline tox.ini]: Why is this line necessary?

File: vsts_ci/darwin/continuous-build-darwin.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      curl -Lo /Users/vsts/.docker/machine/cache/boot2docker.iso https://github.com/boot2docker/boot2docker/releases/download/v18.06.1-ce/boot2docker.iso
+      docker-machine create default --driver xhyve --xhyve-boot2docker-url /Users/vsts/.docker/machine/cache/boot2docker.iso
+      docker-machine env default
+      eval $(docker-machine env default)
+      brew services start docker-machine
+      brew install docker-compose
+      docker version
+      sudo -E tox -e "py$(python.version)"
+      docker-machine rm default -f -y
+    displayName: "Install docker and run tests against iotedgehubdev source code with tox"
```
Comment (LazarusX) [inline vsts_ci/darwin/continuous-build-darwin.yml]: `iotedgehubdev` -> `iotedgedev`

File: vsts_ci/win32/continuous-build-win32.yml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      $env:Path += ";C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\wbin"
+      $env:JAVA_HOME = "C:\Program Files\Java\jdk-11.0.1"
+      choco install python2
+      choco install python --version 3.6.6
+      choco install python --version 3.7.1
+      $env:Path += ";C:\Python37;C:\Python37\Scripts"
+      python -m pip install --upgrade pip
+      pip install tox
+      tox
+    displayName: "Run tests against iotedgehubdev source code"
```
Comment (LazarusX) [inline vsts_ci/win32/continuous-build-win32.yml]: `iotedgehubdev` -> `iotedgedev`

File: tests/test_simulator.py (hunk 2 above)
Comment (neil-yechenwei) [reply tests/test_simulator.py]: Updated code.

File: tox.ini [hunk 6]
```diff
@@ -12,15 +12,17 @@ envlist = py27, py36
 #commands=flake8 iotedgedev
 
 [testenv]
-deps = pytest
+deps = 
+    pytest
+    -rrequirements_dev.txt
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply tox.ini]: If we don't add install required tools, test cases will fail.

File: tox.ini (hunk 3 above)
Comment (neil-yechenwei) [reply tox.ini]: Because for e2e test cases need more variable. Currently more variables are missed in tox env. So I pass all variables to tox env to fix this issue.

File: vsts_ci/darwin/continuous-build-darwin.yml (hunk 4 above)
Comment (neil-yechenwei) [reply vsts_ci/darwin/continuous-build-darwin.yml]: updated code .

File: vsts_ci/win32/continuous-build-win32.yml (hunk 5 above)
Comment (neil-yechenwei) [reply vsts_ci/win32/continuous-build-win32.yml]: updated code .

File: vsts_ci/.vsts-ci.yml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    steps:
+      - template: win32/continuous-build-win32.yml
+
+  - job: MacOS
+    pool:
+      vmImage: macOS-10.13
+    strategy:
+      matrix:
+        Python27:
+          python.version: "2.7"
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply vsts_ci/.vsts-ci.yml]: Yes. we need these to build separated python env in vm.

[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply vsts_ci/.vsts-ci.yml]: python matrix in yml is for enabled cached the python in vm. if not enable this, tox can not find the python while creating virutal env. So I need these code.

File: tox.ini [hunk 8]
```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = py27, py36
+envlist = py2.7, py3.6, py3.7
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply tox.ini]: Updated code to change py2.7 to py27. It seems that it doesn't support the pattern like py2.7

File: vsts_ci/win32/continuous-build-win32.yml [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+      az --version
+    displayName: "Install Azure Cli And Azure Cli Extension"
+
+  - powershell: |
+      $env:Path += ";C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\wbin"
+      $env:JAVA_HOME = "C:\Program Files\Java\jdk-11.0.1"
+      choco install python2
+      choco install python --version 3.6.6
+      choco install python --version 3.7.1
+      $env:Path += ";C:\Python37;C:\Python37\Scripts;C:\Python36;C:\Python36\Scripts;C:\Python27;C:\Python27\Scripts"
```
Comment (LazarusX) [inline vsts_ci/win32/continuous-build-win32.yml]: I think only a Python version is needed for `$PATH`, and since you are appending them to the end of existing `$PATH`, if the existing `$PATH` already contains a Python executable, your change will not be effective at all.

File: vsts_ci/win32/continuous-build-win32.yml [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+      ls C:\
```
Comment (LazarusX) [inline vsts_ci/win32/continuous-build-win32.yml]: Remove this.

File: tests/test_iotedgedev.py [hunk 11]
```diff
@@ -125,11 +129,17 @@ def test_module_add():
     add_module_and_verify(cli.main, runner, "c")
     add_module_and_verify(cli.main, runner, "csharp")
     add_module_and_verify(cli.main, runner, "java")
-    add_module_and_verify(cli.main, runner, "nodejs")
+
+    if (platform.system().lower() == 'windows'):
```
Comment (LazarusX) [inline tests/test_iotedgedev.py]: Please add link to the GitHub issue as comment so that following maintainer knows why adding Node.js modules is skipped on non-Windows.

**Comment by LazarusX** [comment]
Please add a new test with pre-multi-arch deployment manifest template (i.e., without `"$schema-template": "1.0.0"`), so that we can ensure iotedgedev can successfully build and run legacy solutions. Only amd64 platform is needed.

File: vsts_ci/darwin/continuous-build-darwin.yml [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+      sudo chmod u+s $(brew --prefix)/opt/docker-machine-driver-xhyve/bin/docker-machine-driver-xhyve
+      mkdir -p /Users/vsts/.docker/machine/cache
+      curl -Lo /Users/vsts/.docker/machine/cache/boot2docker.iso https://github.com/boot2docker/boot2docker/releases/download/v18.06.1-ce/boot2docker.iso
+      docker-machine create default --driver xhyve --xhyve-boot2docker-url /Users/vsts/.docker/machine/cache/boot2docker.iso
+      docker-machine env default
+      eval $(docker-machine env default)
+      brew services start docker-machine
+      brew install docker-compose
+      docker version
+      sudo -E tox -e "$(TOX_PY_VERSION)"
```
Comment (LazarusX) [inline vsts_ci/darwin/continuous-build-darwin.yml]: Question, with those `py2.7` -> `py27` changes, tox sill cannot find alternative Python on non-Windows?

**Comment by neil-yechenwei** [comment]
> 
> 
> Please add a new test with pre-multi-arch deployment manifest template (i.e., without `"$schema-template": "1.0.0"`), so that we can ensure iotedgedev can successfully build and run legacy solutions. Only amd64 platform is needed.

This PR is only to enable existing test case. I will add this new test case in next pr. Thanks.

File: vsts_ci/win32/continuous-build-win32.yml (hunk 9 above)
Comment (neil-yechenwei) [reply vsts_ci/win32/continuous-build-win32.yml]: Updated code

File: tests/test_iotedgedev.py (hunk 11 above)
Comment (neil-yechenwei) [reply tests/test_iotedgedev.py]: Updated code

File: vsts_ci/darwin/continuous-build-darwin.yml (hunk 12 above)
Comment (neil-yechenwei) [reply vsts_ci/darwin/continuous-build-darwin.yml]: Yes. It still cannot find alternative python on non-windows

File: vsts_ci/darwin/continuous-build-darwin.yml [hunk 13]
```diff
@@ -0,0 +1,59 @@
+steps:
+  - powershell: |
+      ((Get-Content -path "$(BUILD.REPOSITORY.LOCALPATH)\.env.tmp" -Raw) -replace 'IOTHUB_CONNECTION_STRING=".*"','IOTHUB_CONNECTION_STRING="$(IOTHUB_CONNECTION_STRING)"' -replace 'DEVICE_CONNECTION_STRING=".*"','DEVICE_CONNECTION_STRING="$(DEVICE_CONNECTION_STRING)"' -replace 'CONTAINER_REGISTRY_SERVER=".*"','CONTAINER_REGISTRY_SERVER="$(CONTAINER_REGISTRY_SERVER)"' -replace 'CONTAINER_REGISTRY_USERNAME=".*"','CONTAINER_REGISTRY_USERNAME="$(CONTAINER_REGISTRY_USERNAME)"' -replace 'CONTAINER_REGISTRY_PASSWORD=".*"','CONTAINER_REGISTRY_PASSWORD="$(CONTAINER_REGISTRY_PASSWORD)"') | Set-Content -Path "$(BUILD.REPOSITORY.LOCALPATH)\.env.tmp"
```
Comment (adashen) [inline vsts_ci/darwin/continuous-build-darwin.yml]: It seems the env setup steps could be shared between darwin/linux/windows. Can we extract it as a separate step?

Comment (neil-yechenwei) [reply vsts_ci/darwin/continuous-build-darwin.yml]: It seems that it cannot extract a common step to job level.
